2020/day-18/arora-aditya/day18.py

```diff
@@ -0,0 +1,112 @@
+from typing import List, Dict, Set
+from collections import *
+from functools import lru_cache
+from pprint import pprint as pp
+import sys
+from math import *
+from helper.submit.submit import *
+from utils import *
+
+
+DAY = 18
+setup(DAY)
+lines = read_file(DAY)
+
+
+# Part 1
+##################################################
+def split(part):
+    k = []
+    for p in part.split("+"):
+        k.append(p.split("*"))
+
+def get_matching_brackets(line):
+    di = {}
+    st = []
+    L = len(line)
+    for i in range(L):
+        l = line[i]
+        if l == "(":
+            st.append(i)
+        elif l == ")":
+            di[st.pop()] = i
+    if len(st) != 0:
+        sys.exit(1)
+    # print(st, di)
+    return di
+    
+
+def get_ans(line):
+    # print("L", line)
+    split = line.split(" ")
+    bs = get_matching_brackets(line)
+    # print("B", bs)
+    L = len(line)
+    S = len(split)
+    ops = {"+", "*"}
+    j = 0
+    i = 0
+    a = 0
+    O = ""
+    while i < S:
+        s = split[i]
+        if s.startswith("("):
+            # print("H", s, j, bs)
+            att = get_ans(line[j+1:bs[j]])
+            if O == "":
+                a = att
+            elif O == "+":
+                a += att
+            elif O == "*":
+                a *= att
+            i += line[j+1:bs[j]].count(" ") + 1
+            j += len(line[j+1:bs[j]]) + 3
+        elif s in ops:
+            # print("O", s)
+            j += 2
+            i += 1
+            O = s
+        else:
+            k = int(s)
+            j += len(s) + 1
+            # print("N", k)
+            i += 1
+            if O == "":
+                a = k
+            elif O == "+":
+                a += k
+            elif O == "*":
+                a *= k
+    return a
+            
+        
+
+def part1(lines: List[str]) -> int:
+    ans = sum(map(get_ans, lines))
+    return ans
+
+submit(1, part1(lines))
+
+
+# Part 2
+##################################################
+class new_ops(int):
+    def __mul__(self, b):
+        return new_ops(int(self) + b)
+    def __sub__(self, b):
+        return new_ops(int(self) * b)
+    
+    
+def get_ans2(line):
+    e = re.sub(r"(\d+)", r"a(\1)", line)
+    e = e.replace("*", "-")
+    e = e.replace("+", "*")
+    return eval(e, {}, {"a": new_ops})
+
+def part2(lines: List[str]) -> int:
+    ans = sum(map(get_ans2, lines))
+    return ans
```

Comment:
I died parsing as I did for part 1 for this _even_ though it essentially a smaller change (compared to regular math) than part 1

finally hacked this shit together tho

Reply:
Yaaaa today would not have been good if I didn't find some parser code online. I need to work through this in a more clean way, I want to try making a lexer/parser for it, then evaluating recursively. But check out @VictorLi5611's [solution](https://github.com/CarletonComputerScienceSociety/advent-of-code-2020/blob/a1887ecebe5f95140de4de5e541c4d3efe38709d/2020/day-18/VictorLi5611/day18.py), it's quite elegant and makes use of how Python has everything as a class